pycbc/vetoes/sgchisq.py

```diff
@@ -0,0 +1,97 @@
+from pycbc.waveform.utils import apply_fseries_time_shift
+from pycbc.filter import sigma            
+from pycbc.waveform import sinegauss
+from pycbc.vetoes.chisq import SingleDetPowerChisq
+import numpy
+
+class SingleDetSGChisq(SingleDetPowerChisq):
+    """Class that handles precomputation and memory management for efficiently
+    running the power chisq in a single detector inspiral analysis.
+    """
+    def __init__(self, bank, num_bins=0,
+                       snr_threshold=None,
+                       chisq_locations=None):
+        if snr_threshold:
+            self.do = True
+            self.num_bins = num_bins
+        else:
+            self.do = False
+        
+        self.snr_threshold = snr_threshold
+        self.params = {}
+        for descr in chisq_locations:
+            region, values = descr.split(":")
+            mask = bank.table.parse_boolargs([(1, region), (0, 'else')])[0]
+            hashes = bank.table['template_hash'][mask.astype(bool)]
+            for h in hashes:
+                self.params[h] = values
+
+    def values(self, stilde, template, psd, snrv, snr_norm, indices):
+        """ Calculate sine-Gaussian chisq
+
+        Returns
+        -------
+        chisq: Array
+            Chisq values, one for each sample index
+        """
+        if template.params.template_hash not in self.params:
+            return numpy.ones(len(snrv))
+        
+        values = self.params[template.params.template_hash].split(',')
+        if not self.do:
+            return None
+        
+        # Get the chisq bins to use as the frequency reference point
+        bins = self.cached_chisq_bins(template, psd)
+
+        # This is implemented slowly, so let's not call it often, OK?
+        chisq = numpy.ones(len(snrv))
+        for i in range(len(snrv)):
+            snr = abs(snrv[i] * snr_norm)
+            if snr < self.snr_threshold:
+                continue
+
+            N = (len(template) - 1) * 2
+            dt = 1.0 / (N * template.delta_f)
+            kmin = int(template.f_lower / psd.delta_f)
+            time = float(template.epoch) + dt * indices[i]
+            # Shift the time of interest to be centered on 0
+            stilde_shift = apply_fseries_time_shift(stilde, -time)
+
+            qwindow = 50
+            chisq[i] = 0
+            fpeak = (bins[-2] * 2.0 - bins[-3]) * template.delta_f
+            fstop = len(template) * template.delta_f * 0.9
```

Comment:
i dont know what is being done here either.  indices -2 and -3 seem hard coded without further explanation